runtime/src/iree/base/internal/math.h

```diff
@@ -336,9 +336,30 @@ static inline float iree_math_make_f32_from_bits(uint32_t src, int exp_bits,
   } else if (src_exp == 0 && src_mantissa == 0) {
     // Zero. Leave f32_exp and f32_mantissa as zero.
   } else {
-    f32_exp = ((src_exp >> src_exp_shift) + f32_exp_bias - src_exp_bias)
-              << f32_exp_shift;
-    f32_mantissa = src_mantissa << (f32_mantissa_bits - src_mantissa_bits);
+    if (src_exp == 0) {
+      // Source is denormal.
+
+      // Case when the source and destination types have the same exponent
+      // range (bfloat16/f32).
+      if (src_exp_bits == f32_exp_bits) {
+        f32_exp = src_exp << (f32_exp_shift - src_exp_shift);
+        f32_mantissa = src_mantissa << (f32_mantissa_bits - src_mantissa_bits);
+      } else {
+        // Source is denormal, destination is normal.
+        int src_arithmetic_exp = 1 - src_exp_bias;
+        int f32_arithmetic_exp = -src_exp_bias;
+        f32_exp = (f32_arithmetic_exp + f32_exp_bias) << f32_exp_shift;
+        f32_mantissa =
+            (src_mantissa << -(-f32_mantissa_bits + src_mantissa_bits +
```

Comment:
Could you spell out the math here a bit? I'm trying to work out why this is correct and I think I'm missing something obvious

Reply:
Done, let me know if it makes more sense.